libs/wire-api-federation/src/Wire/API/Federation/Version.hs

```diff
@@ -83,108 +90,57 @@ versionInfo = VersionInfo (toList supportedVersions)
-versionPairSchema =
-  object "VersionPair" $
-    (,)
-      <$> fst .= field "from" schema
-      <*> snd .= field "until" schema
-
-data VersionRange
-  = AllVersions
-  | -- | The version in the argument represent an inclusive bound.
-    FromVersion Version
-  | -- | The version in the argument represent an exclusive bound.
-    UntilVersion Version
-  | -- | The second argument represents an exclusive upper bound.
-    FromUntilVersion Version Version
+data VersionRange = VersionRange
+  { _fromVersion :: Maybe Version,
+    _toVersionExcl :: Maybe Version
+  }
 
 deriving instance Eq VersionRange
 
 deriving instance Show VersionRange
 
 deriving instance Ord VersionRange
 
-makePrisms ''VersionRange
+makeLenses ''VersionRange
 
 instance ToSchema VersionRange where
   schema =
     object "VersionRange" $
-      fromTagged
-        <$> toTagged
-          .= bind
-            (fst .= field "tag" versionRangeTagSchema)
-            (snd .= fieldOver _1 "value" untaggedSchema)
+      VersionRange
+        <$> _fromVersion .= maybe_ (optField "from" schema)
+        <*> _toVersionExcl .= maybe_ (optFieldWithDocModifier "until_excl" desc schema)
     where
-      toTagged :: VersionRange -> (VersionRangeTag, VersionRange)
-      toTagged d@AllVersions = (VersionRangeTagAll, d)
-      toTagged d@(FromVersion _) = (VersionRangeTagFrom, d)
-      toTagged d@(UntilVersion _) = (VersionRangeTagUntil, d)
-      toTagged d@(FromUntilVersion _ _) = (VersionRangeTagFromUntil, d)
-
-      fromTagged :: (VersionRangeTag, VersionRange) -> VersionRange
-      fromTagged = snd
-
-      untaggedSchema = dispatch $ \case
-        VersionRangeTagAll -> tag _AllVersions null_
-        VersionRangeTagFrom -> tag _FromVersion (unnamed schema)
-        VersionRangeTagUntil -> tag _UntilVersion (unnamed schema)
-        VersionRangeTagFromUntil -> tag _FromUntilVersion $ unnamed versionPairSchema
+      desc = description ?~ "exlusive upper version bound"
 
 deriving via Schema VersionRange instance ToJSON VersionRange
 
 deriving via Schema VersionRange instance FromJSON VersionRange
 
--- | Compute the lower and upper boundary of a version range. The first
--- component of the pair is the lower boundary, while the second component is
--- the upper boundary. The upper boundary is inclusive.
-versionRangeToBoundaries :: VersionRange -> (Version, Version)
-versionRangeToBoundaries AllVersions = (minBound @Version, maxBound @Version)
-versionRangeToBoundaries (FromVersion fv) = (fv, maxBound @Version)
-versionRangeToBoundaries (UntilVersion uv) = (minBound @Version, pred uv)
-versionRangeToBoundaries (FromUntilVersion fv uv) = (fv, pred uv)
-
--- | Checks if a version is within a given version range.
-inVersionRange :: Version -> VersionRange -> Bool
-inVersionRange v vr =
-  let (lo, hi) = versionRangeToBoundaries vr
-   in lo <= v && v < hi
+allVersions :: VersionRange
+allVersions = VersionRange Nothing Nothing
```

Comment:
You haven't solved that problem you've brought up, namely that you can define one value in two ways, right? This is the same semantically as `VersionRange (Just V0) Nothing`.

Reply:
also good catch!  [fixed](https://github.com/wireapp/wire-server/pull/3831/commits/601d298b00a987caaf77f9cd1dabcab3aca0612b)!